### DISCUSSION TITLE
史上最完整的 markdown 教學，不信你看完還不會 markdown

### DISCUSSION THREAD

**Discussion by 1chooo** [1101-GS4719-C/Python-Community-GS4719-C#2015] [General]
# How to Markdown       

Before [MARKDOWN EVERYTHING], you need to know how to markdown. This document will help you to understand the basic markdown syntax.

## Headers

To use the headers, you can use `#` to `######` for Header 1 to Header 6. And it will be helpful to understand the structure of the document.

# This is Header 1      
###### This is Header 6 

<img width="955" alt="header" src="https://github.com/NCU-GS4538-Linux/Linux-Community/assets/94162591/6ef39361-18df-411b-8aac-0d7025d4c83a">

```markdown
# This is Header 1
## This is Header 2
### This is Header 3
#### This is Header 4
##### This is Header 5
###### This is Header 6
```

## Emphasis

To make the words vividly, you can add the emphasis adaptively.

- *This text will be italic* or _This text will be italic_
- **This text will be bold** or __This text will be bold__
- ***This text will be italic and bold*** or ___This text will be italic and bold___
- ~~This text will be strikethrough~~
- <u>This text will be underlined</u>
- <sup>This text will be superscript</sup>
- <sub>This text will be subscript</sub>

<img width="825" alt="emphasis" src="https://github.com/NCU-GS4538-Linux/Linux-Community/assets/94162591/7bead700-5b42-4a00-bc72-8192743a7422">

```markdown
- *This text will be italic* or _This text will be italic_
- **This text will be bold** or __This text will be bold__
- ***This text will be italic and bold*** or ___This text will be italic and bold___
- ~~This text will be strikethrough~~
- <u>This text will be underlined</u>
- <sup>This text will be superscript</sup>
- <sub>This text will be subscript</sub>
```

## Lists

There are two types of lists, unordered and ordered. You can use `*` for unordered and `1.` for ordered. And you can use the combination of them.

### Unordered

- `* Item 1`
- `* Item 2`
  - `* Item 2a`
  - `* Item 2b`
  - `* Item 2c`
    - `* Item 2c1`
    - `* Item 2c2`

<img width="738" alt="unordered" src="https://github.com/NCU-GS4538-Linux/Linux-Community/assets/94162591/c0fd0099-551a-4538-b3e5-ffddd973a0f8">

```markdown
- `* Item 1`
- `* Item 2`
  - `* Item 2a`
  - `* Item 2b`
  - `* Item 2c`
    - `* Item 2c1`
    - `* Item 2c2`
```

### Ordered

1. `1. Item 1`
2. `2. Item 2`
   1. `1. Item 2a`
   2. `2. Item 2b`
   3. `3. Item 2c`
      1. `1. Item 2c1`
      2. `2. Item 2c2`

<img width="738" alt="ordered" src="https://github.com/NCU-GS4538-Linux/Linux-Community/assets/94162591/29b7203a-fbf8-4e69-9f47-e9401d03f206">

```markdown
1. `1. Item 1`
2. `2. Item 2`
   1. `1. Item 2a`
   2. `2. Item 2b`
   3. `3. Item 2c`
      1. `1. Item 2c1`
      2. `2. Item 2c2`
```

### Combined

1. `1. Item 1`
2. `2. Item 2`
   - `* Item 2a`
   - `* Item 2b`
   - `* Item 2c`
      1. `1. Item 2c1`
      2. `2. Item 2c2`

<img width="738" alt="combined" src="https://github.com/NCU-GS4538-Linux/Linux-Community/assets/94162591/6cb20708-148f-4f01-8ab9-3ece2b7c9805">

```markdown
1. `1. Item 1`
2. `2. Item 2`
   - `* Item 2a`
   - `* Item 2b`
   - `* Item 2c`
      1. `1. Item 2c1`
      2. `2. Item 2c2`
```

> [!TIP]
>
> To insert a line break within a text field, you can use the keyboard shortcut `Shift + Enter`. This allows you to start a new line without ending the current paragraph.

## Links

We can add links in markdown with the hyerlink.

- [Hugo's Portfolio](https://1chooo.com)
- [Hugo's Blog](https://blog.1chooo.com)

<img width="737" alt="links" src="https://github.com/NCU-GS4538-Linux/Linux-Community/assets/94162591/2135681e-8894-44ff-bbbf-747ad7d06d89">

```markdown
- [Hugo's Portfolio](https://1chooo.com)
- [Hugo's Blog](https://blog.1chooo.com)

Sample: [alt text](url)
```

## Images

We can embed images in markdown with the file path or URL.

<img width="1392" alt="demo" src="https://github.com/NCU-GS4538-Linux/Linux-Community/assets/94162591/1b3efccc-0366-4bc0-b629-fa577ee47a60">

<img width="1008" alt="image" src="https://github.com/NCU-GS4538-Linux/Linux-Community/assets/94162591/7e7e558c-329e-4bfe-a7ed-bac78e6b7cc2">

```markdown
![Markdown Everything](./demo.png)

Sample: ![alt text](url)
```

## Code

We can use inline code and code block in markdown to represent the code.

### Inline Code

We can use the backticks <code>\`</code> to create the inline code.

`print("Hello, World!")`

<img width="875" alt="inline-code" src="https://github.com/NCU-GS4538-Linux/Linux-Community/assets/94162591/1b5444a8-6c0f-4dc1-a1d4-8b00179e63d0">

```markdown
`print("Hello, World!")`
```

### Code Block

With three backticks <code>\`\`\`</code>, we can create the code block. Also, we can specify the language to highlight the syntax next to the first three backticks, such as <code>\`\`\`python</code>.

```python
def foo() -> str:
    """
    This is a function to return the string 'bar'
    
    Returns:
        str: 'bar'
    """
    return 'bar'
```

<img width="821" alt="code-block" src="https://github.com/NCU-GS4538-Linux/Linux-Community/assets/94162591/5870418d-e6ec-475a-828a-ee347c580d29">

<pre><code class=language-markdown>```python
def foo() -> str:
    """
    This is a function to return the string 'bar'
    
    Returns:
        str: 'bar'
    """
    return 'bar'
```
</code></pre>


## Quotes

We can use the `>` to create the quotes in markdown.

> "This is a quote."
>
> 1. Item 1
> 2. Item 2
>     - Item 2a

> **Python Hello World:**
>
> ```python
> print('Hello, World!')
> ```

<img width="773" alt="quotes" src="https://github.com/NCU-GS4538-Linux/Linux-Community/assets/94162591/b44c96f5-973a-4934-8fe3-278863c97240">



```markdown
> "This is a quote."
>
> 1. Item 1
> 2. Item 2
>     - Item 2a

> **Python Hello World:**
>
> ```python
> print('Hello, World!')
> ```
```

## Alert

In GitHub, it supports the render of alerts. You can use the following syntax to create alerts, also known as callouts.

> [!NOTE]
> Useful information that users should know, even when skimming content.

> [!TIP]
> Helpful advice for doing things better or more easily.

> [!IMPORTANT]
> Key information users need to know to achieve their goal.

> [!WARNING]
> Urgent info that needs immediate user attention to avoid problems.

> [!CAUTION]
> Advises about risks or negative outcomes of certain actions.

<img width="992" alt="alert" src="https://github.com/NCU-GS4538-Linux/Linux-Community/assets/94162591/567afd55-f385-4cd0-a44a-8ca08895b525">

```markdown
> [!NOTE]
> Useful information that users should know, even when skimming content.

> [!TIP]
> Helpful advice for doing things better or more easily.

> [!IMPORTANT]
> Key information users need to know to achieve their goal.

> [!WARNING]
> Urgent info that needs immediate user attention to avoid problems.

> [!CAUTION]
> Advises about risks or negative outcomes of certain actions.
```

## Tables

You can create the tables to represent the data in markdown.

| Header 1 | Header 2 | Header 3 |
| -------- | -------- | -------- |
| Cell 1   | Cell 2   | Cell 3   |
| Cell 4   | Cell 5   | Cell 6   |

<img width="773" alt="table" src="https://github.com/NCU-GS4538-Linux/Linux-Community/assets/94162591/94e5ac2d-cbbb-4334-9df6-bc531518fec2">

```markdown
| Header 1 | Header 2 | Header 3 |
| -------- | -------- | -------- |
| Cell 1   | Cell 2   | Cell 3   |
| Cell 4   | Cell 5   | Cell 6   |
```

## Ignore Markdown

If you want to ignore the markdown syntax, you can use the backslash `\`. If you want to ignore the markdown syntax in the whole line, you can use the `<!-- -->` to comment out the line.

Let's \*\*rename\*\* the variable \`var\` in with \`var2\` in the \`main.py\`

<img width="745" alt="ignore-md" src="https://github.com/NCU-GS4538-Linux/Linux-Community/assets/94162591/82d1239e-449b-4998-9781-dc0a0abd2da2">

```markdown
Let's \*\*rename\*\* the variable \`var\` in with \`var2\` in the \`main.py\`

<!-- This content will not appear in the rendered Markdown -->
```

> [!TIP]
>
> Sometimes, I will use `HTML` to ignore the markdown syntax.

## Footnotes

We can add the footnotes in markdown to provide additional information.

This is Hugo Speaking 👨🏻‍💻 [^1], [^bignote], [^2]

[^1]: This is the first footnote.

[^2]: [Hugo](https://1chooo.com)

[^bignote]: Here's one with multiple paragraphs and code.

    Indent paragraphs to include them in the footnote.

    `{ my code }`

    Add as many paragraphs as you like.

    ```python
    def foo():
        return 'bar'
    ```

> [!TIP]
>
> You can install [Markdown-Everything] [^4] to improve the rendering of the markdown in the [VSCode].

<img width="857" alt="footnotes" src="https://github.com/NCU-GS4538-Linux/Linux-Community/assets/94162591/6bae2598-9b1d-47d6-a42f-12dba87443e9">

[^4]: [1chooo/Markdown-Everything](https://github.com/1chooo/markdown-everything)

[Markdown-Everything]: https://marketplace.visualstudio.com/items?itemName=1chooo.markdown-everything

[VSCode]: https://code.visualstudio.com/

```markdown
This is Hugo Speaking 👨🏻‍💻 [^1], [^bignote], [^2]

[^1]: This is the first footnote.

[^2]: [Hugo](https://1chooo.com)

[^bignote]: Here's one with multiple paragraphs and code.

    Indent paragraphs to include them in the footnote.

    `{ my code }`

    Add as many paragraphs as you like.

    ```python
    def foo():
        return 'bar'
    ```
```

## Math

In markdown, we can use the $\LaTeX$ syntax to represent the math equations. We can use the inline math with `$` and block math with `$$`.

- Inline Math: $E = mc^2$
- Block Math:

$$
\int_{-\infty}^\infty e^{-x^2} dx = \sqrt{\pi}
$$

- Algorithm:

$$
\begin{array}{l}
\textbf{Algorithm 1 QuickSort(array A, l, r)} \\
1.\ \text{if}\ r - l + 1 \leq 1\ \text{then} \\
2.\ \quad \text{return} \\
3.\ \text{end if} \\
4.\ p = \text{ChoicePivot}(A, l, r) \\
5.\ \text{Partition}(A, p, l, r) \\
6.\ \text{QuickSort}(A, l, p - 1) \\
7.\ \text{QuickSort}(A, p + 1, r) \\
\end{array}
$$

<img width="1008" alt="math" src="https://github.com/NCU-GS4538-Linux/Linux-Community/assets/94162591/7eac2529-fe19-4a62-83dc-d1fb83e07538">



```markdown
- Inline Math: $E = mc^2$
- Block Math:

$$
\int_{-\infty}^\infty e^{-x^2} dx = \sqrt{\pi}
$$

- Algorithm:

$$
\begin{array}{l}
\textbf{Algorithm 1 QuickSort(array A, l, r)} \\
1.\ \text{if}\ r - l + 1 \leq 1\ \text{then} \\
2.\ \quad \text{return} \\
3.\ \text{end if} \\
4.\ p = \text{ChoicePivot}(A, l, r) \\
5.\ \text{Partition}(A, p, l, r) \\
6.\ \text{QuickSort}(A, l, p - 1) \\
7.\ \text{QuickSort}(A, p + 1, r) \\
\end{array}
$$
```

## Expandable Sections

We can create the expandable sections in markdown to hide the content.

<details>

<summary>👉🏻 Click Me to expand</summary>

Hi.

<img width="1392" alt="demo" src="https://github.com/NCU-GS4538-Linux/Linux-Community/assets/94162591/960d6981-aab8-4a5e-a12c-31117b3502d5">

</details>

<details>

<summary>👉🏻 Click Me to expand</summary>

Hi.

<img width="1392" alt="demo" src="https://github.com/NCU-GS4538-Linux/Linux-Community/assets/94162591/960d6981-aab8-4a5e-a12c-31117b3502d5">

</details>

<img width="1136" alt="expandable-sections" src="https://github.com/NCU-GS4538-Linux/Linux-Community/assets/94162591/72a1d33b-ecc0-4e92-b168-5fba2e98b99f">

```markdown
<details>

<summary>👉🏻 Click Me to expand</summary>

Hi.

![](./demo.png)
</details>

<details>

<summary>👉🏻 Click Me to expand</summary>

Hi.

![](./demo.png)
</details>
```

## Chart

We can plot a variety of charts with `mermaid` syntax in markdown. Let's take `flowchart` as an example.

```mermaid
---
title: Multistep Processing of a User Program by Hugo
---
flowchart-elk LR
    A((Source Program)) --> B[Compiler or Assembler]
    B[Compiler or Assembler] --> C((Object Module))
    C((Object Module)) --> E[Linkage Editor]
    D(("`Other 
    Object 
    Modules`")) --> E[Linkage Editor]
    H((System Library)) --> G[Loader]
    G[Loader] --> J[in-memory binary memory image]
    I(("`Dynamically 
    Loaded 
    System 
    Library`")) --> J[in-memory binary memory image]

    subgraph Compile Time
        direction TB
        B[Compiler or Assembler]
    end
    subgraph Load Time
        direction TB
        E[Linkage Editor] --> F((Load Module))
        F((Load Module)) --> G[Loader]
    end
    subgraph Execution Time
        direction TB
        J[in-memory binary memory image]
    end
```

<img width="1062" alt="flowchart" src="https://github.com/NCU-GS4538-Linux/Linux-Community/assets/94162591/bc8570cd-1ecd-4486-9f2c-169f256ede92">

<pre><code class=language-markdown>```markdown
---
title: Multistep Processing of a User Program by Hugo
---
flowchart-elk LR
    A((Source Program)) --> B[Compiler or Assembler]
    B[Compiler or Assembler] --> C((Object Module))
    C((Object Module)) --> E[Linkage Editor]
    D(("`Other 
    Object 
    Modules`")) --> E[Linkage Editor]
    H((System Library)) --> G[Loader]
    G[Loader] --> J[in-memory binary memory image]
    I(("`Dynamically 
    Loaded 
    System 
    Library`")) --> J[in-memory binary memory image]

    subgraph Compile Time
        direction TB
        B[Compiler or Assembler]
    end
    subgraph Load Time
        direction TB
        E[Linkage Editor] --> F((Load Module))
        F((Load Module)) --> G[Loader]
    end
    subgraph Execution Time
        direction TB
        J[in-memory binary memory image]
    end
```
</code></pre>

[MARKDOWN EVERYTHING]: https://marketplace.visualstudio.com/items?itemName=1chooo.markdown-everything

**Comment by 1chooo** [comment]
> [!IMPORTANT]
> 本文教學的原因是出自我寫了一個 VSCode 的 Extension 👉🏻 [Markdown Everything](https://marketplace.visualstudio.com/items?itemName=1chooo.markdown-everything)，大家可以下載來增強 VSCode 的 Markdown 的撰寫體驗，同時也可以前往 [Markdown Everything - how2md](https://github.com/1chooo/markdown-everything/blob/main/docs/how2md.md) 查看原文。

**Comment by 710388** [comment]
看完以後還是覺得不會，是不是我太笨了

**Comment by OHoln** [comment]
Markdown真的是學幾次忘幾次，每次學完過了一段時間沒用下次要寫又是各種google，果然要學好的關鍵曬要常常使用，加強大腦的神經連結

**Comment by 155Julia** [comment]
感謝版主的整理，我是markdown的常用者，但卻還是對它像陌生人一樣，每次會用的語法就那簡單的幾個，因此寫出來的筆記都偏醜，但這次有了版主的整理看來可以試試看新的花招了。
